File: app/src/main/java/com/orgzly/android/ui/Selection.java
package com.orgzly.android.ui;

import android.content.res.TypedArray;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.view.View;

import com.orgzly.R;

import java.util.Collection;
import java.util.TreeSet;

public class Selection {
    private static final String SAVED_BUNDLE_KEY = "list_of_selected_ids";

    /** IDs of selected notes. */
    private TreeSet<Long> mSelectedIds = new TreeSet<>();

    @ColorInt
    private int bgColor = 0;

    public void updateView(View view, long noteId) {
        if (bgColor == 0) {
            TypedArray arr = view.getContext().obtainStyledAttributes(R.styleable.ColorScheme);
            bgColor = arr.getColor(R.styleable.ColorScheme_item_selected_bg_color, 0);
            arr.recycle();
        }

        if (mSelectedIds.contains(noteId)) {
            view.setBackgroundColor(bgColor);
        } else {
            view.setBackgroundResource(0);
        }
    }

    public int getCount() {
        return mSelectedIds.size();
    }

    public TreeSet<Long> getIds() {
        return mSelectedIds;
    }

    public boolean contains(long noteId) {
        return mSelectedIds.contains(noteId);
    }

    public void select(long noteId) {
        mSelectedIds.add(noteId);
    }

    public void deselect(long noteId) {
        mSelectedIds.remove(noteId);
    }

    public void deselectAll(Collection<Long> noteIds) {
        mSelectedIds.removeAll(noteIds);
    }

    public void toggle(View view, long noteId) {
        if (contains(noteId)) {
            deselect(noteId);
        } else {
            select(noteId);
        }

        if (view != null) {
            updateView(view, noteId);
        }
    }

    public void clearSelection() {
        mSelectedIds.clear();
    }

    /**
     * Save selected items.
     * Restored with {@link Selection#restoreIds(android.os.Bundle)}.
     */
    public void saveIds(Bundle bundle) {
        if (getCount() > 0) {
            long[] idsArray = new long[mSelectedIds.size()];
            int i = 0;
            for (long id: mSelectedIds) {
                idsArray[i++] = id;
            }
            bundle.putLongArray(SAVED_BUNDLE_KEY, idsArray);
        } else {
            bundle.remove(SAVED_BUNDLE_KEY);
        }
    }

    /**
     * Restore selected items.
     * Saved with {@link Selection#saveIds(android.os.Bundle)}.
     */
    public void restoreIds(Bundle bundle) {
        mSelectedIds.clear();

        if (bundle != null && bundle.containsKey(SAVED_BUNDLE_KEY)) {
            long[] ids = bundle.getLongArray(SAVED_BUNDLE_KEY);
            if (ids != null) {
                for (long id : ids) {
                    mSelectedIds.add(id);
                }
            }
        }
    }
}

File: app/src/main/java/com/orgzly/android/ui/Selection.kt
package com.orgzly.android.ui

import android.os.Bundle
import androidx.annotation.ColorInt
import android.view.View
import com.orgzly.R
import com.orgzly.android.ui.util.styledAttributes
import java.util.*

class Selection {
    private val idSet = TreeSet<Long>()

    /** Map from adapter ID to real note ID, used for agenda. */
    private var idMap: HashMap<Long, Long>? = null

    fun getIds(): Set<Long> {
        return (idMap?.let { map -> idSet.mapNotNull { map[it] } } ?: idSet).toSet()
    }

    fun getOnly(): Long? {
        return if (idSet.isNotEmpty()) {
            getIds().first()
        } else {
            null
        }
    }

    val count: Int
        get() = idSet.size


    operator fun contains(id: Long): Boolean {
        return idSet.contains(id)
    }

    fun toggle(id: Long) {
        if (contains(id)) {
            idSet.remove(id)
        } else {
            idSet.add(id)
        }
    }

    // TODO: Choose mode when creating selection
    fun toggleSingleSelect(id: Long) {
        if (contains(id)) {
            idSet.clear()
        } else {
            idSet.clear()
            idSet.add(id)
        }
    }

    fun clear() {
        idSet.clear()
    }

    fun setMap(map: HashMap<Long, Long>) {
        idMap = map
    }

    fun removeNonExistent(dataIds: Set<Long>) {
        idSet.removeAll { it !in dataIds }
    }

    /**
     * Save selected items.
     * Restored with [Selection.restoreIds].
     */
    fun saveIds(bundle: Bundle) {
        if (count > 0) {
            val idsArray = LongArray(idSet.size)
            var i = 0
            for (id in idSet) {
                idsArray[i++] = id
            }
            bundle.putLongArray(SAVED_BUNDLE_KEY, idsArray)
        } else {
            bundle.remove(SAVED_BUNDLE_KEY)
        }
    }

    /**
     * Restore selected items.
     * Saved with [Selection.saveIds].
     */
    fun restoreIds(bundle: Bundle?) {
        idSet.clear()

        if (bundle != null && bundle.containsKey(SAVED_BUNDLE_KEY)) {
            val ids = bundle.getLongArray(SAVED_BUNDLE_KEY)
            if (ids != null) {
                for (id in ids) {
                    this.idSet.add(id)
                }
            }
        }
    }

    @ColorInt
    private var selectionBgColor = 0

    fun setIsSelectedBackground(view: View, id: Long) {
        if (selectionBgColor == 0) {
            selectionBgColor = view.context.styledAttributes(R.styleable.ColorScheme) { typedArray ->
                typedArray.getColor(R.styleable.ColorScheme_item_selected_bg_color, 0)
            }
        }

        if (contains(id)) {
            view.setBackgroundColor(selectionBgColor)
        } else {
            view.setBackgroundResource(0)
        }
    }

    companion object {
        private const val SAVED_BUNDLE_KEY = "list_of_selected_ids"
    }
}
